src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -543,10 +543,23 @@ private void RaiseProjectDiagnosticsCreated(Project project, StateSet stateSet,
 
             foreach (var documentId in newAnalysisResult.DocumentIds)
             {
-                // TryGetSourceGeneratedDocumentForAlreadyGeneratedId is safe to use here because the only way to report
-                // a diagnostic in a source generated document is for the source generated document to exist.
-                var document = project.GetTextDocument(documentId)
-                    ?? project.TryGetSourceGeneratedDocumentForAlreadyGeneratedId(documentId);
+                var document = project.GetTextDocument(documentId);
+
+                // If we couldn't find a normal document, and all features are enabled for source generated documents,
+                // attempt to locate a matching source generated document in the project.
+                if (document is null
+                    && project.Solution.Workspace.Services.GetService<ISyntaxTreeConfigurationService>() is { EnableOpeningSourceGeneratedFilesInWorkspace: true })
+                {
+                    if (_workspaceThreadingService is not null)
+                    {
+                        document = _workspaceThreadingService.Run(() => project.GetSourceGeneratedDocumentAsync(documentId, CancellationToken.None).AsTask());
+                    }
+                    else
+                    {
+                        document = project.GetSourceGeneratedDocumentAsync(documentId, CancellationToken.None).AsTask().WaitAndGetResult_CanCallOnBackground(CancellationToken.None);
+                    }
```

Comment:
Does this get called on the background always or was there a concern this could run on the UI thread? I'm unsure if you're doing this pattern just to follow the standard pattern or if indeed this does run on the UI thread (which seems surprising, generally.)

Reply:
➡️ Implemented just to follow the standard pattern (JTF-aware but not required)